LibraBFT/ImplShared/Util/RWST.agda

```diff
@@ -5,175 +5,236 @@
+-- outputs are concatenated. The postcondition desired for the above sequence
+-- becomes a postcondition for `f x` in which the outputs of `m` are prepended
+-- to the outputs of `f x`.
+RWST-Post++ : ∀ {Wr St A} → RWST-Post Wr St A → List Wr → RWST-Post Wr St A
+RWST-Post++ P outs x post outs₁ = P x post (outs ++ outs₁)
+
+-- Consider again the sequence `RWST-bind m₁ λ x → f x`. We also translate a
+-- postcondition `P` for the sequence into a postcondition for `m` ---
+-- specifically, the post condition we want for `m` is that the weakest
+-- precondition for `RWST-Post++ P outs` holds, where `outs` are the outputs of
+-- `m`.
+RWST-weakestPre-bindPost  : (ev : Ev) (f : A → RWST Ev Wr St B) → RWST-Post Wr St B → RWST-Post Wr St A
+RWST-weakestPre-ebindPost : (ev : Ev) (f : A → RWST Ev Wr St (C ⊎ B)) → RWST-Post Wr St (C ⊎ B) → RWST-Post Wr St (C ⊎ A)
+
+RWST-weakestPre : (m : RWST Ev Wr St A) → RWST-PredTrans Ev Wr St A
+RWST-weakestPre (RWST-return x) P ev pre = P x pre []
+RWST-weakestPre (RWST-bind m f) P ev pre =
+  RWST-weakestPre m (RWST-weakestPre-bindPost ev f P) ev pre
+RWST-weakestPre (RWST-gets f) P ev pre = P (f pre) pre []
+RWST-weakestPre (RWST-put post) P ev pre = P unit post []
+RWST-weakestPre RWST-ask P ev pre = P ev pre []
+RWST-weakestPre (RWST-tell outs) P ev pre = P unit pre outs
+RWST-weakestPre (RWST-if (clause (b ≔ c) gs)) P ev pre =
+  (toBool b ≡ true → RWST-weakestPre c P ev pre)
+  × (toBool b ≡ false → RWST-weakestPre (RWST-if gs) P ev pre)
+RWST-weakestPre (RWST-if (otherwise≔ c)) P ev pre =
+  RWST-weakestPre c P ev pre
+RWST-weakestPre (RWST-either e f₁ f₂) P ev pre =
+    (∀ x → (e ≡ Left x) →
+      RWST-weakestPre (f₁ x) P ev pre)
+  × (∀ y → (e ≡ Right y) →
+       RWST-weakestPre (f₂ y) P ev pre)
+RWST-weakestPre (RWST-ebind m f) P ev pre =
+  RWST-weakestPre m (RWST-weakestPre-ebindPost ev f P) ev pre
+RWST-weakestPre (RWST-maybe m f₁ f₂) P ev pre =
+  (m ≡ nothing → RWST-weakestPre f₁ P ev pre)
+  × (∀ j → m ≡ just j → RWST-weakestPre (f₂ j) P ev pre)
+
+RWST-weakestPre-ebindPost ev f Post (Left r) post outs =
+  Post (Left r) post outs
+RWST-weakestPre-ebindPost ev f Post (Right r) post outs =
+  ∀ c → c ≡ r → RWST-weakestPre (f c) (RWST-Post++ Post outs) ev post
+
+RWST-weakestPre-bindPost ev f Post x post outs =
+  ∀ r → r ≡ x → RWST-weakestPre (f r) (RWST-Post++ Post outs) ev post
+
+-- For every RWST computation `m`, `RWST-Contract m` is the type of proofs that,
+-- for all post conditions `P`, starting environments `ev` and prestates `pre`,
+-- to prove that `P` holds after running `m` in `ev` and `pre`, it suffices to
+-- provide the weakest precondition for `P` with respect to `m`, `ev`, and `pre`.
+RWST-Contract : (m : RWST Ev Wr St A) → Set₁
+RWST-Contract{Ev}{Wr}{St}{A} m =
+  (P : RWST-Post Wr St A)
+  → (ev : Ev) (pre : St) → RWST-weakestPre m P ev pre
+  → let (x , post , outs) = RWST-run m ev pre in
+    P x post outs
+
+-- This proves `RWST-Contract m` for arbitrary `m`.
```

Comment:
It would be helpful to add something like: This proves that our definition of `RWST-weakestPre` provides a _sufficient_ precondition for the post condition to hold.  It does _not_ prove that it is the weakest such precondition (even though it's true); this is not important for our purposes.